node/impl/client/client.go

```diff
@@ -483,84 +466,103 @@ func (a *API) makeRetrievalQuery(ctx context.Context, rp rm.RetrievalPeer, paylo
+	if err != nil {
+		return nil, xerrors.Errorf("failed to create temp CARv2 file: %w", err)
+	}
+	// make sure to remove the CARv2 file if anything goes wrong from here on.
+	defer func() {
+		if finalErr != nil {
+			_ = os.Remove(carV2File)
+		}
+	}()
+
+	var root cid.Cid
+	if ref.IsCAR {
+		root, err = transformCarToCARv2(ref.Path, carV2File)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to import CAR file: %w", err)
+		}
+	} else {
+		root, err = a.importNormalFileToCARv2(ctx, id, ref.Path, carV2File)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to import normal file to CARv2: %w", err)
+		}
+	}
+
 	if err := a.imgr().AddLabel(id, importmgr.LSource, "import"); err != nil {
 		return nil, err
 	}
-
 	if err := a.imgr().AddLabel(id, importmgr.LFileName, ref.Path); err != nil {
 		return nil, err
 	}
-
-	nd, err := a.clientImport(ctx, ref, st)
-	if err != nil {
+	if err := a.imgr().AddLabel(id, importmgr.LCARv2FilePath, carV2File); err != nil {
 		return nil, err
 	}
-
-	if err := a.imgr().AddLabel(id, importmgr.LRootCid, nd.String()); err != nil {
+	if err := a.imgr().AddLabel(id, importmgr.LRootCid, root.String()); err != nil {
 		return nil, err
 	}
 
 	return &api.ImportRes{
-		Root:     nd,
+		Root:     root,
 		ImportID: id,
 	}, nil
 }
 
-func (a *API) ClientRemoveImport(ctx context.Context, importID multistore.StoreID) error {
-	return a.imgr().Remove(importID)
-}
-
-func (a *API) ClientImportLocal(ctx context.Context, f io.Reader) (cid.Cid, error) {
-	file := files.NewReaderFile(f)
-
-	id, st, err := a.imgr().NewStore()
+func (a *API) ClientRemoveImport(ctx context.Context, importID uint64) error {
+	info, err := a.imgr().Info(importID)
 	if err != nil {
-		return cid.Undef, err
+		return xerrors.Errorf("failed to fetch import info: %w", err)
 	}
-	if err := a.imgr().AddLabel(id, "source", "import-local"); err != nil {
-		return cid.Cid{}, err
+
+	// remove the CARv2 file if we've created one.
+	if path := info.Labels[importmgr.LCARv2FilePath]; path != "" {
+		_ = os.Remove(path)
 	}
 
-	bufferedDS := ipld.NewBufferedDAG(ctx, st.DAG)
+	return a.imgr().Remove(importID)
+}
 
-	prefix, err := merkledag.PrefixForCidVersion(1)
+func (a *API) ClientImportLocal(ctx context.Context, r io.Reader) (cid.Cid, error) {
+	// write payload to temp file
+	tmpPath, err := a.imgr().NewTempFile(rand.Uint64())
 	if err != nil {
 		return cid.Undef, err
 	}
-	prefix.MhType = DefaultHashFunction
-
-	params := ihelper.DagBuilderParams{
-		Maxlinks:  build.UnixfsLinksPerLevel,
-		RawLeaves: true,
-		CidBuilder: cidutil.InlineBuilder{
-			Builder: prefix,
-			Limit:   126,
-		},
-		Dagserv: bufferedDS,
-	}
-
-	db, err := params.New(chunker.NewSizeSplitter(file, int64(build.UnixfsChunkSize)))
+	defer os.Remove(tmpPath) //nolint:errcheck
+	tmpF, err := os.Open(tmpPath)
 	if err != nil {
 		return cid.Undef, err
 	}
-	nd, err := balanced.Layout(db)
-	if err != nil {
+	defer tmpF.Close() //nolint:errcheck
```

Comment:
I don't think we need this `defer tmpF.Close()` as we call `tmpF.Close()` immediately below

Reply:
The copy statement before that could fail.